crud/common/rebalance.lua

```diff
@@ -0,0 +1,211 @@
+        hook()
+    end
+
+    log.info('Rebalance safe mode disabled')
+end
+
+local function rebalance_init()
+    M.metrics.enable_storage_metrics()
+
+    -- box.watch was introduced in tarantool 2.10.0
+    if not utils.tarantool_supports_box_watch() then
+        log.warn('This version of tarantool does not support autoswitch to safe mode during rebalance. '
+            .. 'Update to newer version or use `_crud.rebalance_safe_mode_enable()` to enable safe mode manually.')
+        return
+    end
+
+    box.watch('box.status', function()
+        if box.info.ro then
+            return
+        end
+
+        local stored_safe_mode
+        if box.space[SETTINGS_SPACE_NAME] == nil then
+            create_space()
+            box.space[SETTINGS_SPACE_NAME]:insert{ 'safe_mode', false }
+        else
+            stored_safe_mode = box.space[SETTINGS_SPACE_NAME]:get{ 'safe_mode' }
+        end
+        M.safe_mode = stored_safe_mode and stored_safe_mode.value or false
+
+        if M.safe_mode then
+            for hook, _ in pairs(M.safe_mode_enable_hooks) do
+                hook()
+            end
+        else
+            box.space._bucket:on_replace(safe_mode_trigger)
+            for hook, _ in pairs(M.safe_mode_disable_hooks) do
+                hook()
+            end
+        end
+    end)
+
+    box.watch(SAFE_MOD_ENABLE_EVENT, function(_, do_enable)
+        if box.info.ro or not do_enable then
+            return
+        end
+        safe_mode_enable()
+    end)
+end
+
+local function router_cache_clear()
+    M._router_cache_last_clear_ts = fiber.time()
+    return utils.get_vshard_router_instance():_route_map_clear()
+end
+
+local function router_cache_length()
+    return utils.get_vshard_router_instance().known_bucket_count
+end
+
+local function router_cache_last_clear_ts()
+    return M._router_cache_last_clear_ts
+end
+
+-- Rebalance related metrics
+local function enable_storage_metrics()
+    if not has_metrics_module then
+        return
+    end
+
+    local safe_mode_enabled_gauge = metrics.gauge(
+            'tnt_crud_storage_safe_mode_enabled',
+            "is safe mode enabled on this storage instance"
+    )
+
+    metrics.register_callback(function()
+        safe_mode_enabled_gauge:set(safe_mode_status() and 1 or 0)
+    end)
+end
+
+local function enable_router_metrics()
+    if not has_metrics_module then
+        return
+    end
+
+    local router_cache_length_gauge = metrics.gauge(
+            'tnt_crud_router_cache_length',
+            "number of bucket routes in vshard router cache"
+    )
+    local router_cache_last_clear_ts_gauge = metrics.gauge(
+            'tnt_crud_router_cache_last_clear_ts',
+            "when vshard router cache was cleared last time"
+    )
+
+    metrics.register_callback(function()
+        router_cache_length_gauge:set(router_cache_length())
+        router_cache_last_clear_ts_gauge:set(router_cache_last_clear_ts())
+    end)
+end
+
+M.init = rebalance_init
```

Comment:
Nit: let's make the code consistent with all other modules, it seems we tend to make the variable with proper name and assign methods to it `.`:

https://github.com/tarantool/crud/blob/c52a6f588bdadc04d5b3e9f438d78851aa7e8055/crud/stats/local_registry.lua#L39-L39